### ISSUE TITLE
test failure in stackage nightly: "preserve pipe behaviour FAILED"

### ISSUE THREAD

**Issue by juhp** [yesodweb/Shelly.hs#106]
One of shelly's tests seems to have started failing after Stackage Nightly 2015-07-28.

```
Selbstverständlich überraschend
  with binary handle mode
  can detect failing commands in pipes
bar
  preserve pipe behaviour FAILED [1]

Failures:

  1) bash preserve pipe behaviour
       expected: 0
        but got: 141

Randomized with seed 1132532841

Finished in 0.1296 seconds
51 examples, 1 failure
Test suite shelly-testsuite: FAIL
Test suite logged to:
/home/stackage/work/logs/nightly/shelly-1.6.3.1/test-run.out
0 of 1 test suites (0 of 1 test cases) passed.
```

**Comment by adinapoli** [comment]
I suspect this is another ripple effect of https://github.com/yesodweb/Shelly.hs/issues/103 and it seems to be system specific (I wasn't able to reproduce this on my system). Perhaps it's worthwhile investigating why it's happening?

@juhp on which machine those tests are run? (in terms of OS version etc).

**Comment by juhp** [comment]
The Stackage builder is running Ubuntu 14.04.1 LTS.

Is it possible that it is only intermittent?

**Comment by adinapoli** [comment]
@juhp The only relevant(ish) thing I was able to find was this SO question:

http://stackoverflow.com/questions/22464786/ignoring-bash-pipefail-for-error-code-141

It mention the SIGPIPE error which can yield error code 141 if the pipe no longer exist, but I cannot see how this can happen. Might it be related on the particular way the Stackage tests are run?

**Comment by gregwebs** [comment]
@adinapoli can you add something to trap and ignore sigpipe? Also, how about setting `errExit` according to whether it is set in shelly?

**Comment by adinapoli** [comment]
Hey Greg, I'm on holiday for a full week without access to a Haskell
compiler ;) I would like to dig deeper into this and on why it's failing,
only intermittently, on Linux, but perhaps the quickest solution for now
would be to simply mark the test as pending until my return, unless of
course you want to perform the investigation yourself. Your call ;)

Alfredo

**Comment by gregwebs** [comment]
also reported on the Nix build in #108

**Comment by adinapoli** [comment]
Ah, that's very annoying! Apologies, I didn't have time to look into it so far (I might have a bit of calm next week).

**Comment by juhp** [comment]
Hmm dunno I am still getting a failure in stackage:

```
:
bar
  preserve pipe behaviour FAILED [1]

Failures:

  test/src/RunSpec.hs:49:
  1) bash preserve pipe behaviour
       expected: 0
        but got: 141

Randomized with seed 457505560

Finished in 0.2832 seconds
51 examples, 1 failure
Test suite shelly-testsuite: FAIL
Test suite logged to: /home/stackage/work/logs/nightly/shelly-1.6.3.4/test-run.out
```

**Comment by gregwebs** [comment]
I removed the pipefail feature @adinapoli you can add it back in your code by using `bashSetPipefail`. So this should be fixed in 1.6.4

**Comment by adinapoli** [comment]
Sounds good to me. Shame this little function created so many problems!
